src/Microsoft.AspNetCore.Razor.Language/RazorProjectEngine.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Razor.Language
+{
+    public abstract class RazorProjectEngine
+    {
+        public abstract RazorProjectFileSystem ProjectFileSystem { get; }
+
+        public abstract RazorEngine Engine { get; }
+
+        public abstract IReadOnlyList<IRazorProjectEngineFeature> Features { get; }
+
+        public abstract RazorCodeDocument Process(string filePath);
+
+        public abstract RazorCodeDocument Process(RazorSourceDocument sourceDocument);
+
+        public static RazorProjectEngine Create(RazorProjectFileSystem projectFileSystem) => Create(projectFileSystem, configure: null);
+
+        public static RazorProjectEngine Create(RazorProjectFileSystem projectFileSystem, Action<RazorProjectEngineBuilder> configure)
+        {
+            if (projectFileSystem == null)
+            {
+                throw new ArgumentNullException(nameof(projectFileSystem));
+            }
+
+            var builder = new DefaultRazorProjectEngineBuilder(designTime: false, projectFileSystem: projectFileSystem);
+
+            AddRuntimeDefaults(builder);
+            configure?.Invoke(builder);
+
+            return builder.Build();
+        }
+
+        public static RazorProjectEngine CreateDesignTime(RazorProjectFileSystem projectFileSystem) => CreateDesignTime(projectFileSystem, configure: null);
+
+        public static RazorProjectEngine CreateDesignTime(RazorProjectFileSystem projectFileSystem, Action<RazorProjectEngineBuilder> configure)
+        {
+            if (projectFileSystem == null)
+            {
+                throw new ArgumentNullException(nameof(projectFileSystem));
+            }
+
+            var builder = new DefaultRazorProjectEngineBuilder(designTime: true, projectFileSystem: projectFileSystem);
+
+            AddDesignTimeDefaults(builder);
+            configure?.Invoke(builder);
+
+            return builder.Build();
+        }
+
+        private static void AddDesignTimeDefaults(RazorProjectEngineBuilder builder)
+        {
+            var defaultEngine = RazorEngine.CreateDesignTime();
```

Comment:
Yes, we're creating a dummy engine. This is valuable because we support creating `RazorEngine`s independently of a project engine; otherwise we'd have to add a new project engine specific call path to the RazorEngine creation code.

Ultimately we throw away the engine shell and take all of its features/phases.

Reply:
That's not valid. Features aren't expected to support reinitialization (setting `Engine` twice).

I want to understand more about why you think this is necessary.

Reply:
> That's not valid. Features aren't expected to support reinitialization (setting Engine twice).

Good point. Gonna have to think about what to do here.



> I want to understand more about why you think this is necessary.

When configuring the Razor system we want to show all `Engine` features to users who are customizing the `IRazorFeature` collection.